[WORK_IN_PROGRESS] Blob storage

elenasolomon opened the pull request:
Implementation for:
- create container
- list containers
- create block blob (not finished), data block blob (which will store JSON files)
- list blobs
- update the content of a data block blob
- remove blob
- remove container

What I want to do next:
- better documentation
- more unit tests
- support large files (size more than 256MB)
- finish the BlockBlob implementation
- support for AppendBlob and AppendDataBlob
- authentication with SAS

elenasolomon committed:
small refactoring

elenasolomon committed:
draft implementation

elenasolomon committed:
Implementation for:
- create container
- create DataBlockBlob
- update DataBlockBlob
- list containers
- list blobs
- remove container
- remove blob

elenasolomon committed:
- changes the way the blobs are named in unit tests

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+    options = options || {};
+    assert(options, 'options must be specified.');
+    assert(options.container, 'The container instance, `options.container`, must be specified.');
+    assert(options.container instanceof Container, '`options.container` must be an instance of Container.');
```
I'm not sure that we want to use the `instanceof` operator for our own defined type.  [Here's an example of this failing](http://stackoverflow.com/questions/18261788/nodejs-require-instanceof-behaviour).  Instead, let's assume that if a container is passed in that it's a valid one.  If we wanted to be extra careful, we could check for the presence of an important method or two on the container object.  Maybe `assert(typeof options.container.createDataBlob === 'function')`

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+
+    this.name = options.name;
+    this.type = options.type;
+    this.eTag = options.eTag;
```
What's the reason for allowing the eTag to be set in the constructor?  Shouldn't we only be setting it after hitting the azure endpoints?

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+    options.blobType = this.type;
+
+    try {
+      await this.blobService.putBlob(this.container.name, this.name, options, content);
```
does the response for the putBlob method return the eTag?  If so, how about we update the eTag of the Blob here?

jhford replied:
aha, looking at the `create()` method above, it seems that it does return the eTag.  Let's store it here.  Maybe we could also store an in-memory reference to the content as `this.content`?

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+   * Load the content of the blob
+   */
+  async load() {
+    let blob = await this.blobService.getBlob(this.container.name, this.name);
```
Should we use conditional headers here to ensure that we only `getBlob` only if the eTag of our local data doesn't match the copy on the server?

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+      if (!valid) {
+        debug(`Failed to validate the blob content against schema with id: 
+            ${this.container.schemaId}, errors: ${this.container.validate.errors}`);
+        throw new BlobDoesNotConformToSchema(`Failed to validate the blob content against schema with id: 
```
Can we rename this?  `SchemaValidationError` is a little easier to read.  I'm not sure if it's a standard, but I try to end all my custom errors with `Error`.

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+        version: 1,
+      });
+    } catch (error) {
+      debug(`Failed to serialize the content of the blob: ${this.name}`);
```
can you dump the serialization error to the screen as well as the name of the blob?

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+    let data = await super.load();
+    let content = JSON.parse(data).content;
+    // Validate the JSON against the schema
+    this._validateJSON(content);
```
In the case where the validation fails, I'd like to still be able to get at the data.  How about catching the exception and throwing a new exception which has a `content` property as well as the failures in validation attached.

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+   *
+   * @param content - a JSON object
+   */
+  async create(json) {
```
the name here doesn't match the doc.  Let's use the `content` name.

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,252 @@
+          debug('ERROR: MAX_MODIFY_ATTEMPTS exhausted, we might have congestion');
+          throw new CongestionError('MAX_MODIFY_ATTEMPTS exhausted, check for congestion');
+        }
+        return attemptModify();
```
this looks great!  Can we have an exponential backoff implemented here?  Without an exponential backoff we'll end up exacerbating the problem if it really is congestion by hammering the network `MAX_MODIFY_ATTEMPTS` times

jhford commented on src/blobstorage.js:
```diff
@@ -0,0 +1,253 @@
+ * The base name of the JSON schema used created the schema identifier
+ * @const
+ */
+const JSON_SCHEMA_NAME_BASE_URL = 'http://schemas.taskcluster.net/';
```
shouldn't this end up being set to the Azure url?  Since we're in essence publishing the schema at that location, we should be able to use the real Azure URL as the schema's ID right?

jhford commented on src/blobstorage.js:
```diff
@@ -0,0 +1,253 @@
+    let validate;
+    let schemaId = `${JSON_SCHEMA_NAME_BASE_URL}${name}#`;
+    try {
+      let schemaValidation = this.validator.getSchema(schemaId);
```
I think for now we should not load based on the id, unless the id is the one that would match `JSON_SCHEMA_BLOB_NAME`.  Either way is OK for me.

jhford commented on src/container.js:
```diff
@@ -1,67 +1,161 @@
+
+    this.blobService = options.blobService;
+    this.name = options.name;
+    this.eTag = options.eTag;
```
like blob, we should only be setting this value based on the result of an API call.

jhford commented on src/customerrors.js:
```diff
@@ -0,0 +1,32 @@
+class ExtendableError extends Error {
```
I like this as a base class!  Nicely done

jhford reviewed (commented):
This looks really great!  There's a couple of things that should be changed before merging though.  I skimmed the tests, since they're likely to change after updating the PR.  Let me know if you have any questions.

Great work!

elenasolomon committed:
Integration of code review changes (part 1)

elenasolomon committed:
- offer the option to preserve the content of the blob in memory
- a small refactoring for create and load container methods
- removed the eTag from the Container constructor

elenasolomon committed:
Adds the README

jhford commented on README.md:
```diff
@@ -0,0 +1,86 @@
+azure-blob-storage
```
Thanks for writing this!  It looks really good.

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,272 @@
+  }
+
+  _validateJSON(content) {
+    if (content && this.container.schemaId) {
```
when would we want to validate JSON against a schema without a content body to validate against?

elenasolomon replied:
You are right!I will correct the function

jhford commented on src/blob.js:
```diff
@@ -0,0 +1,272 @@
+  }
+
+  async append(options, content) {
+    // 1. validate the content against the schema
```
This is a todo still, right?

elenasolomon replied:
Yes. Everything related with append blob is in progress

jhford commented on src/container.js:
```diff
@@ -1,67 +1,269 @@
+  }
+
+  _getSchemaId() {
+    return `http://${this.blobService.options.accountId}.blob.core.windows.net/${this.name}/${JSON_SCHEMA_BLOB_NAME}#`;
```
even if we don't upload the schema, let's stick with this name for the id

jhford reviewed (approved):
Looks good!
